runtime/lua/vim/shared.lua

```diff
@@ -987,4 +987,17 @@ do
   end
 end
 
+--- Memoize a function {fn} using {hash_fn} to hash the arguments.
+--- Internally uses a weak table to cache the results of {fn}.
+--- @generic F: function
+--- @param fn F Function to memoize.
```

Comment:
w.r.t. https://github.com/neovim/neovim/pull/25154#issuecomment-1719266502, this is tricky since technically `hash_fn` and `fn` can both be functions, but given that `fn` is always a function, it may make most sense to put it last